application_sandbox/synchronization2_write_timestamp/main.cpp

```diff
@@ -0,0 +1,482 @@
+// Copyright 2022 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+
+#include "application_sandbox/sample_application_framework/sample_application.h"
+#include "mathfu/matrix.h"
+#include "mathfu/vector.h"
+#include "support/entry/entry.h"
+#include "vulkan_helpers/buffer_frame_data.h"
+#include "vulkan_helpers/helper_functions.h"
+#include "vulkan_helpers/vulkan_application.h"
+#include "vulkan_helpers/vulkan_model.h"
+
+using Mat44 = mathfu::Matrix<float, 4, 4>;
+using Vector4 = mathfu::Vector<float, 4>;
+using Vector3 = mathfu::Vector<float, 3>;
+namespace torus_model {
+#include "torus_knot.obj.h"
+}
+const auto& torus_data = torus_model::model;
+
+uint32_t torus_vertex_shader[] =
+#include "write_timestamp.vert.spv"
+    ;
+
+uint32_t torus_fragment_shader[] =
+#include "write_timestamp.frag.spv"
+    ;
+
+struct WriteTimestampFrameData {
+  containers::unique_ptr<vulkan::VkCommandBuffer> command_buffer_;
+  containers::unique_ptr<vulkan::VkFramebuffer> framebuffer_;
+  containers::unique_ptr<vulkan::DescriptorSet> torus_descriptor_set_;
+  containers::unique_ptr<vulkan::VkBufferView> timestamp_buf_view_;
+};
+
+VkPhysicalDeviceSynchronization2FeaturesKHR kSynchronization2Features = {
+    VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SYNCHRONIZATION_2_FEATURES_KHR, nullptr,
+    VK_TRUE};
+
+// This creates an application with 16MB of image memory, and defaults
+// for host, and device buffer sizes.
+class WriteTimestampSample
+    : public sample_application::Sample<WriteTimestampFrameData> {
+ public:
+  WriteTimestampSample(const entry::EntryData* data,
+                       const VkPhysicalDeviceFeatures& requested_features)
+      : data_(data),
+        Sample<WriteTimestampFrameData>(
+            data->allocator(), data, 1, 512, 1, 1,
+            sample_application::SampleOptions()
+                .EnableDepthBuffer()
+                .EnableMultisampling()
+                .SetVulkanApiVersion(VK_API_VERSION_1_1)
+                .AddDeviceExtensionStructure(&kSynchronization2Features),
+            requested_features, {}, {VK_KHR_SYNCHRONIZATION_2_EXTENSION_NAME}),
+        torus_(data->allocator(), data->logger(), torus_data),
+        grey_scale_(0u),
+        num_frames_(0u) {
```

Comment:
grey_scale_ and num_frames_ seems not used?

Reply:
Yeah, it looks unused, but probably not causing any harm. That part is just copied from write_timestamp, and I don't want to add unnecessary delta between write_timestamp and synchronization2_write_timestamp.